File: BinaryTreeRightSideView_199.java
import java.util.Queue;
import java.util.*;


//BFS: Using Queue level order traversal. Based on the queue size, we can iterate. In that if we can take only the last element it would be the right side visible element. 
//DFS: Using iterative pre order, Each deapth for the first time we are adding the node to  the result list. In that we only add right side elements.

//TC: O(N) where N is the height of the tree, since we travers through all the elements.
//SC: O(N) - Since we are using Queue for the auxilary purposes. ( can ignore the list as that is what function expects back)
class BinaryTreeRightSideView_993 {
    public List<Integer> rightSideViewBFS(TreeNode root) {
        List<Integer> result = new ArrayList<Integer>(); 
        if(root == null) {
            return result;
        }
        Queue<TreeNode> queue = new LinkedList();
        queue.add(root);
        while(!queue.isEmpty()) {
            int size = queue.size();
            for(int i=0;i<size;i++) {
                TreeNode currentNode = queue.poll();
                if(i == size-1) { // Gets always right element. //1 - 0 th element, 4 - 3rd element.
                    result.add(currentNode.val);
                }
                if(currentNode.left!=null) {
                    queue.add(currentNode.left);
                }
                if(currentNode.right!=null) {
                    queue.add(currentNode.right);
                }
            }
        }

        return result;
    }


    public List<Integer> rightSideViewDFS(TreeNode root) {
        List<Integer> result = new ArrayList<Integer>();
        rightView(root, result, 0);
        return result;
    }
    
    public void rightView(TreeNode curr, List<Integer> result, int currDepth){
        if(curr == null){
            return;
        }
        if(currDepth == result.size()){
            result.add(curr.val);
        }
        
        rightView(curr.right, result, currDepth + 1);
        rightView(curr.left, result, currDepth + 1);
        
    }

    public static void main(String[] args) {
        BinaryTreeRightSideView_993 binaryTreeRightSideView_993 = new BinaryTreeRightSideView_993();
        TreeNode root = new TreeNode().getSampleFour();
        //final List<Integer> result = binaryTreeRightSideView_993.rightSideViewBFS(root);
        final List<Integer> result = binaryTreeRightSideView_993.rightSideViewDFS(root);
        System.out.println("The given nodes are cousins:" + result);
    }
}

File: binary-tree-right-side-view_199.py
#DFS Level order approach and exploring first right child then left
#Time Complexity O(N)
#Space Complexity O(h)
#It successfully runs on leetcode
from collections import deque
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        self.result=[]
        if not root: return []
        self.dfs(root,0)
        return self.result
    def dfs(self,root,level):
            if not root: return
            if level > len(self.result)-1:
                self.result.append(root.val)
            self. dfs(root.right,level+1)
            self.dfs(root.left,level+1)
            
#DFS with left child first and then overriding right most element
from collections import deque
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        self.result=[]
        if not root: return []
        self.dfs(root,0)
        return self.result
    def dfs(self,root,level):
            if not root: return
            if level > len(self.result)-1:
                self.result.append(root.val)
            else:
                self.result[level]=root.val
            self. dfs(root.left,level+1)
            self.dfs(root.right,level+1)


#BFS Level order Approach and then taking the last element of each level
#Time Complexity O(N)
#Space Complexity O(N)
#It successfully runs on leetcode
from collections import deque
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        result=[]
        if not root: return result
        q= deque()
        q.append(root)
        while q:
            l=[]
            for i in range(len(q)):
                cur= q.popleft()
                l.append(cur.val)
                if cur.left: q.append(cur.left)
                if cur.right: q.append(cur.right)
            result.append(l[-1])
        return result


from collections import deque
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        result=[]
        if not root: return result
        q= deque()
        q.append(root)
        size=len(q)
        while q:
            for i in range(size):
                cur= q.popleft()
                if i == size-1:
                    result.append(cur.val)
                if cur.left: q.append(cur.left)
                if cur.right: q.append(cur.right)
            size = len(q)
        return result